### DISCUSSION TITLE
Safer serialization for SQLAlchemy jobstore

### DISCUSSION THREAD

**Discussion by Jaakkonen** [agronholm/apscheduler#710] [General]
Currently SQLAlchemy jobstore seems to be doing plain `pickle.dumps` serialization on the job objects, writing that to a string SQL column and then loading those off from there with `pickle.loads`. This is a bit unsafe since if a malicious actor has access to write to the SQL database they can add a serialized object with overridden methods such as `__getitem__` and `__getattr_` which will be invoked when loading the jobs. https://github.com/agronholm/apscheduler/blob/a2247473fb0f664ce02f6fb0c584a003df1e952b/apscheduler/jobstores/sqlalchemy.py#L130-L148

I'm wondering whether it would be possible to serialize the jobs to some JSON format/etc where SQL write -> arbitrary code execution vulnerabilities would be harder to pull off. Is there a hard dependence on using `pickle` (e.g. necessity to serialize functions) or would it be possible to avoid using pickle?

**Comment by agronholm** [comment]
You mean as described [here](https://apscheduler.readthedocs.io/en/master/userguide.html#using-persistent-data-stores)? Unfortunately 3.x users are out of luck since it won't get any new features.
